tools/clang/unittests/HLSLExec/ExecutionTest_SER.h

```diff
@@ -0,0 +1,2554 @@
+    ++Histo[Val];
+  VERIFY_ARE_EQUAL(Histo.size(), 1);
+  VERIFY_ARE_EQUAL(Histo[0], 4096);
+}
+
+TEST_F(ExecutionTest, SERWaveIncoherentHitTest) {
+  // Test SER with wave incoherent conditional assignment of HitObject values
+  //    with and without procedural attributes.
+  static const char *ShaderSrc = R"(
+struct SceneConstants
+{
+    float4 eye;
+    float4 U;
+    float4 V;
+    float4 W;
+    float sceneScale;
+    uint2 windowSize;
+    int rayFlags;
+};
+
+struct[raypayload] PerRayData
+{
+    uint visited : read(anyhit,closesthit,miss,caller) : write(anyhit,miss,closesthit,caller);
+};
+
+struct Attrs
+{
+    float2 barycentrics : BARYCENTRICS;
+};
+
+struct CustomAttrs
+{
+    float dist;
+};
+
+RWStructuredBuffer<int> testBuffer : register(u0);
+RaytracingAccelerationStructure topObject : register(t0);
+ConstantBuffer<SceneConstants> sceneConstants : register(b0);
+
+RayDesc ComputeRay()
+{
+    uint2   launchIndex = DispatchRaysIndex().xy;
+    uint2   launchDim = DispatchRaysDimensions().xy;
+
+    float2 d = float2(DispatchRaysIndex().xy) / float2(DispatchRaysDimensions().xy) * 2.0f - 1.0f;
+    RayDesc ray;
+    ray.Origin = sceneConstants.eye.xyz;
+    ray.Direction = normalize(d.x*sceneConstants.U.xyz + d.y*sceneConstants.V.xyz + sceneConstants.W.xyz);
+    ray.TMin = 0;
+    ray.TMax = 1e18;
+
+    return ray;
+}
+
+static const uint ProceduralHitKind = 11;
+
+[shader("raygeneration")]
+void raygen()
+{
+    uint2   launchIndex = DispatchRaysIndex().xy;
+    uint2   launchDim = DispatchRaysDimensions().xy;
+
+    RayDesc ray = ComputeRay();
+
+    PerRayData payload;
+    payload.visited = 0;
+
+    dx::HitObject hitObject;
+
+    // Use wave incoherence to decide how to create the HitObject
+    if (launchIndex.x % 4 == 1)
+    {
+        ray.Origin.x += 2.0f;
+        hitObject = dx::HitObject::TraceRay(topObject, RAY_FLAG_SKIP_CLOSEST_HIT_SHADER, 0xFF, 0, 0, 0, ray, payload);
+    }
+    else if (launchIndex.x % 4 == 2)
+    {
+        hitObject = dx::HitObject::TraceRay(topObject, RAY_FLAG_SKIP_PROCEDURAL_PRIMITIVES, 0xFF, 0, 0, 0, ray, payload);
+    }
+    else if (launchIndex.x % 4 == 3)
+    {
+        hitObject = dx::HitObject::TraceRay(topObject, RAY_FLAG_SKIP_TRIANGLES, 0xFF, 0, 0, 0, ray, payload);
+    }
+
+    dx::MaybeReorderThread(hitObject);
+
+    if (hitObject.IsNop())
+        payload.visited |= 1U;
+    if (hitObject.IsMiss())
+        payload.visited |= 2U;
+
+    if (hitObject.GetHitKind() == ProceduralHitKind)
+        payload.visited |= 8U;
+    else if (hitObject.IsHit())
+        payload.visited |= 4U;
+
+    dx::HitObject::Invoke(hitObject, payload);
+
+    // Store the result in the buffer
+    int id = launchIndex.x + launchIndex.y * launchDim.x;
+    testBuffer[id] = payload.visited;
+}
+
+[shader("miss")]
+void miss(inout PerRayData payload)
+{
+    // UNUSED
+}
+
+// Triangles
+[shader("anyhit")]
+void anyhit(inout PerRayData payload, in Attrs attrs)
+{
+    AcceptHitAndEndSearch();
```

Comment:
This test appears to suffer from the same issue I described for `SERShaderTableIndexTest`, where the order of triangle vs procedural hits in the acceleration structure determines the overall bucketing.

Reply:
Hit order issue should be resolved. Keeping this open until confirmed.